lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,371 @@
+        )
+        if relation is None:
+            raise ValueError(
+                "Unable to restore {}: relation {} (id={}) not found.".format(
+                    self, snapshot["relation_name"], snapshot["relation_id"]
+                )
+            )
+        self.relation = relation
+
+
+class VaultAutounsealInitialize(ops.EventBase):
+    """Event emitted when Vault autounseal should be initialized for a new application."""
+
+    def __init__(self, handle: ops.Handle, relation: model.Relation):
+        super().__init__(handle)
+        self.relation = relation
+
+    def snapshot(self) -> Dict[str, Any]:
+        """Return snapshot data that should be persisted."""
+        return dict(
+            super().snapshot(),
+            relation_id=self.relation.id,
+            relation_name=self.relation.name,
+        )
+
+    def restore(self, snapshot: Dict[str, Any]) -> None:
+        """Restore the event from a snapshot."""
+        super().restore(snapshot)
+        relation = self.framework.model.get_relation(
+            snapshot["relation_name"], snapshot["relation_id"]
+        )
+        if relation is None:
+            raise ValueError(
+                f"Unable to restore {self}: relation {snapshot['relation_name']} (id={snapshot['relation_id']}) not found."
+            )
+        self.relation = relation
+
+
+class VaultAutounsealDestroy(ops.EventBase):
+    """Event emitted when Vault autounseal configuration for a relation should be destroyed."""
+
+    def __init__(self, handle: ops.Handle, relation: model.Relation):
+        super().__init__(handle)
+        self.relation = relation
+
+    def snapshot(self) -> Dict[str, Any]:
+        """Return snapshot data that should be persisted."""
+        return dict(
+            super().snapshot(),
+            relation_id=self.relation.id,
+            relation_name=self.relation.name,
+        )
+
+    def restore(self, snapshot: Dict[str, Any]) -> None:
+        """Restore the event from a snapshot."""
+        super().restore(snapshot)
+        relation = self.framework.model.get_relation(
+            snapshot["relation_name"], snapshot["relation_id"]
+        )
+        if relation is None:
+            raise ValueError(
+                f"Unable to restore {self}: relation {snapshot['relation_name']} (id={snapshot['relation_id']}) not found."
+            )
+        self.relation = relation
+
+
+class VaultAutounsealProvidesEvents(ops.ObjectEvents):
+    """Events raised by the vault-autounseal relation on the provider side."""
+
+    vault_autounseal_initialize = ops.EventSource(VaultAutounsealInitialize)
+    vault_autounseal_destroy = ops.EventSource(VaultAutounsealDestroy)
+
+
+class VaultAutounsealRequireEvents(ops.ObjectEvents):
+    """Events raised by the vault-autounseal relation on the requirer side."""
+
+    vault_autounseal_details_ready = ops.EventSource(VaultAutounsealDetailsReadyEvent)
```

Comment:
Do we expect requirers to just use `relation_broken`?

Reply:
That's a good question. I admittedly overlooked this because it's effectively handled by `_configure` anyway the next time an event fires.

I'll add an event for this. Makes sense.

Reply:
https://github.com/canonical/vault-k8s-operator/pull/346/commits/c9675c04e4090f53cf9f9fd62a6b9463875ca6a4